### PR TITLE
[inductor] FX graph cache: Add support for symbolic shapes

### PR DESCRIPTION
**Pull request by masnesral**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #111421

Summary: Add support for caching graphs that have tensor args with symbolic shapes. The high-level appraoch is to serialize guards with the on-disk cached object and validating those guards pass before serving a cached object.

Test Plan: New unit tests

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

File: test/inductor/test_codecache.py
Comment (masnesral) [inline test/inductor/test_codecache.py:200]: What's missing here is validation that guards were added after loading from the cache. Still investigating, but does anyone know if there's a straightforward way to access the guards in a structured way so I can do some validation here?

File: test/inductor/test_codecache.py
Comment (masnesral) [inline test/inductor/test_codecache.py:269]: TODO: I also need to check that guards were added properly after a cache hit.

File: torch/_inductor/codecache.py
Comment (masnesral) [inline torch/_inductor/codecache.py:338]: The prefix character seems to be the existing scheme here to differentiate different kinds of hashes for different types of cached objects. Probably I should use an enum. TODO.

File: torch/_inductor/codecache.py
Comment (masnesral) [inline torch/_inductor/codecache.py:565]: Since we need to access the `example_args`, it was convenient to make the `gm` and `example_args` arguments explicit rather than packaging them up in a list.

File: torch/_inductor/codecache.py
Comment (masnesral) [inline torch/_inductor/codecache.py:592]: This method is helpful for printing all the fields that we include in the hash when computing the cache key. It is not currently called anywhere, however, because I think it's waaaay too verbose even for debug logging. Does anyone know if we have precedent for different log levels?

File: torch/_inductor/codecache.py
Comment (masnesral) [inline torch/_inductor/codecache.py:664]: I kept this implementation under the `FxGraphCache` class, but all the methods are static. I prefer the namespacing of putting these methods in the class, but lemme know if a class of all static methods hurts your eyeballs.

File: torch/_inductor/codecache.py
Comment (masnesral) [inline torch/_inductor/codecache.py:715]: Please comment. I don't know if clearing the whole cache is problematic, e.g., from a performance perspective. I could, for example, introduce a new context manager analogous to `suppress_guards()` above that selectively uses a cached vs. non-cached version of `shape_env.evaluate_expr`, but I thought it was worth asking if that's overkill.

File: torch/_inductor/codecache.py
Comment (masnesral) [inline torch/_inductor/codecache.py:793]: I preferred a ctor here rather than extracting all these fields at the caller.

File: torch/_inductor/compile_fx.py
Comment (masnesral) [inline torch/_inductor/compile_fx.py:329]: See comment above. I changed this to be explicit about the two args (`gm`, `example_inputs`)

File: torch/fx/experimental/symbolic_shapes.py
Comment (masnesral) [inline torch/fx/experimental/symbolic_shapes.py:3588]: The changes here are just about splitting apart the `evaluate_guards_for_args` function into smaller pieces that I can use for the cache impl. Namely, I need separate phases for creating the guards expression and evaluating it (in a different context)

File: torch/_inductor/codecache.py
Comment (masnesral) [inline torch/_inductor/codecache.py:609]: The diff looks weird here. `load_graph` is replaced by `lookup_graph` above and the implementation here is for `save_graph`.  So ignore the red lines I guess.

File: torch/_inductor/codecache.py
Comment (eellison) [inline torch/_inductor/codecache.py:485]: It might be useful in the future to canonicalize symints so that `hash(tensor([s1, s1, s2])) == hash(tensor([s3, s3, s4]))`.. not needed now

File: torch/_inductor/codecache.py
Comment (eellison) [reply torch/_inductor/codecache.py:592]: There is log.info vs log.debug. Debug is more verbose

File: torch/_inductor/codecache.py
Comment (eellison) [reply torch/_inductor/codecache.py:657]: Sometimes people prefer `@classmethod` and `cls.method` for this reason i think. up to you.

File: torch/_inductor/codecache.py
Comment (eellison) [reply torch/_inductor/codecache.py:722]: Seems off to me that we would clear cache. leaving this to @ezyang to comment

File: torch/_inductor/codecache.py
Comment (eellison) [inline torch/_inductor/codecache.py:758]: do we need FileLock here ?

**Review by eellison — COMMENTED** [review]
looks good but let's wait for full review from @ezyang

File: torch/_inductor/codecache.py
Comment (masnesral) [reply torch/_inductor/codecache.py:592]: Yeah, I'm saying that even log.debug is probably too verbose here. I think you only want to print this level of detail if you need to deeply debug a cache miss that you think shouldn't be a cache miss and you want to see all the gory details of what went into the key calculation.

Comment (eellison) [reply torch/_inductor/codecache.py:592]: You could enable it with an environment variable

File: torch/_inductor/codecache.py
Comment (Chillee) [inline torch/_inductor/codecache.py:640]: Is it possible to end up loading a more generic graph from cache than we need?

File: torch/_inductor/codecache.py
Comment (Chillee) [reply torch/_inductor/codecache.py:722]: I agree it seems kinda weird that we would need to clear the cache. If a cache hit is preventing you from adding a guard, then that should mean that that guard was already evaluated, no?

File: test/inductor/test_codecache.py
Comment (ani300) [inline test/inductor/test_codecache.py:130]: Same as previous comment

File: torch/_inductor/codecache.py
Comment (masnesral) [reply torch/_inductor/codecache.py:722]: See some of the comments above. The approach here is to:
1. Load a possible match from the cache.
2. Evaluate the guards, but in a mode that _does not_ modify the guards in the current environment, because: in the case that there's a miss, we don't actually want the current env to change.
3. If there's a hit, only then re-evaluate to cause the guards to be added.

The problem here is that the caching at `evaluate_expr()` interferes with #3

File: torch/_inductor/codecache.py
Comment (masnesral) [reply torch/_inductor/codecache.py:640]: Ohhhh, yes. That's a nice catch. So really I need to consider all versions and pick the "best" option, don't I? Hmm, is there any obvious criteria I can use to determine whether one option is more appropriate than the other? cc @eellison

Comment (eellison) [reply torch/_inductor/codecache.py:640]: I think maybe as a follow up - not needed in initial pr imo. Generally the behavior for symbolic shapes is to reuse a compilation if it works.

File: test/inductor/test_codecache.py
Comment (ani300) [inline test/inductor/test_codecache.py:179]: this is equivalent to dynamic=True

**Review by ani300 — COMMENTED** [review]
I made some minor comments on the unit tests wrt the behavior of the `dynamic` parameter. Other than that, not much more to add as @eellison and @Chillee already covered much of the logic questions I had. I defer to @ezyang to see some not-as-obvious issue with the treatment of symbolic shapes, as I'm def not the expert there.

File: torch/fx/experimental/symbolic_shapes.py
Comment (ezyang) [reply torch/fx/experimental/symbolic_shapes.py:3588]: Seems like a good comment to put into the code comments

File: test/inductor/test_codecache.py
Comment (ezyang) [inline test/inductor/test_codecache.py:196]: do we have an API for resetting the on disk compiled graph cache?

Comment (ezyang) [inline test/inductor/test_codecache.py:196]: are you resetting the counters on each test iteration? It would be easier to understand and refactor the tests if you do

File: test/inductor/test_codecache.py
Comment (ezyang) [inline test/inductor/test_codecache.py:233]: So I'm thinking about the test you wrote here, and it seems to me that many pre-existing tests that are checking cache behavior could be repurposed for the FX graph cache. The general idea is that as long as we aren't testing for guards produced by Dynamo, the test can be run in two ways: once without clearing the in memory code cache, and once clearing the in memory code cache, so you hit the FX graph cache. Something to think about.

File: torch/_inductor/codecache.py
Comment (ezyang) [inline torch/_inductor/codecache.py:331]: Why are we truncating the hash?

File: torch/_inductor/codecache.py
Comment (ezyang) [reply torch/_inductor/codecache.py:592]: Don't do an environment variable. Instead, you can create an artifact logger that isn't triggered even in debug. bytecode is done this way, check it out.

File: torch/_inductor/codecache.py
Comment (ezyang) [reply torch/_inductor/codecache.py:640]: Suppose you have a cache for `(s0, 4)` and another for `(5, s1)`. There is no unique best choice cache entry to load.

Probably the only way to actually do this is some sort of autotuning based thing, where for any given concrete size you've benchmarked which one runs fastest, and you load that particular one.

File: torch/_inductor/codecache.py
Comment (ezyang) [inline torch/_inductor/codecache.py:690]: This function needs more logging.

Comment (ezyang) [reply torch/_inductor/codecache.py:690]: In particular, eventually we're going to be debugging weird production issues related to cache related problems, and I need 100% of the relevant file system state we probe to be reported by the logs here, because it's going to be extremely inconvenient to actually get to the machine that has the problem

File: torch/_inductor/codecache.py
Comment (ezyang) [inline torch/_inductor/codecache.py:695]: Let's sort the listdir so that it is deterministic

File: torch/_inductor/codecache.py
Comment (ezyang) [inline torch/_inductor/codecache.py:705]: Is this guaranteed to be in the correct order?

File: torch/_inductor/codecache.py
Comment (ezyang) [reply torch/_inductor/codecache.py:485]: It's not clear to me that this is sound, given https://github.com/pytorch/pytorch/pull/111421/files#diff-c9b517f8db609ffa866804dfa2689188a4fee20abacaa0b0dca91625c1b5cb8dR705

if you say [s0, s1] == [s4, s3], you have to make sure that you know how to flip the SymInt arguments when eval'ing the guards. That sounds difficult.

File: torch/_inductor/codecache.py
Comment (ezyang) [inline torch/_inductor/codecache.py:711]: You don't need `definitely_true` here, just `bool` it. definitely_true is for dealing with unbacked SymInt

File: torch/_inductor/codecache.py
Comment (ezyang) [reply torch/_inductor/codecache.py:722]: @masnesral I think a better solution would be to customize the caching on `evaluate_expr` so that it respects whether or not guards were suppressed or not. WDYT?

**Comment by ezyang** [comment]
The basic logic looks OK. I'm a little nervous about just slamming `filter_symints` but I think it'll be OK since we are hashing on the FX graph.

File: test/inductor/test_codecache.py
Comment (masnesral) [reply test/inductor/test_codecache.py:196]: We don't, but I could add one. Do you have a use case in mind where we'd want to wire it up, or you just want to make it available? There are other caches in this file and the clear() methods don't remove on-disk artifacts. Would we want to add those? Would we want to an API to just blow away all artifacts (independent of type)?

Comment (masnesral) [reply test/inductor/test_codecache.py:196]: Each individual test resets the counters. Is that what you mean? Or you were expecting resets within this individual test body?

Comment (ezyang) [reply test/inductor/test_codecache.py:196]: The testing API probably looks a bit different; instead of blowing away the full cache, you probably just want to cache into some hermetic directory separate from everything else.

File: test/inductor/test_codecache.py
Comment (masnesral) [reply test/inductor/test_codecache.py:200]: @ezyang do you have a recommendation? If I have a torch.compiled function, is there a straightforward way to see that our guards were properly added even in the case of a cache hit?

File: test/inductor/test_codecache.py
Comment (masnesral) [reply test/inductor/test_codecache.py:233]: There is no in-memory FX graph cache in the implementation so far though. It's only on disk. Do you mean clearing on the on-disk cache? (btw, I have a TODO to investigate whether extending to an in-memory cache too would help compile time, but my guess is that it will be in the noise)

File: torch/_inductor/codecache.py
Comment (masnesral) [reply torch/_inductor/codecache.py:338]: I was sticking with the convention already started, but can definitely turn it into a a more descriptive string. There's also an existing precedent for adding an extension, e.g., "cg" which I guessed means "code graph"?

File: torch/_inductor/codecache.py
Comment (masnesral) [reply torch/_inductor/codecache.py:592]: @ezyang.  Sry, I don't know what bytecode refers to. Is this the sample you have in mind? https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/convert_frame.py#L85

File: torch/_inductor/codecache.py
Comment (masnesral) [reply torch/_inductor/codecache.py:690]: @ezyang it's related to the discussion above about using an env var. I assumed that dumping all this info is going to waaay too verbose for normal debug output -- because it's a lot. Lemme know if that assumption is wrong and I'll just log.debug it. I haven't investigated what you mean by using an artifact logger, but does artifact logging the info give you the control you need?

File: torch/_inductor/codecache.py
Comment (masnesral) [reply torch/_inductor/codecache.py:711]: thanks!

File: test/inductor/test_codecache.py
Comment (masnesral) [reply test/inductor/test_codecache.py:89]: So it turns out there's a shortcoming with that change. None seems to be equivalent to False in terms of the FX graph that gets generated (at least for these tests). So I'd need to rearchitect the tests slightly to make sure each test gets a clean tmp directory. I can do that, but I sorta liked the current behavior of leaving the tmp directory intact for the run of the full set of tests because we'd catch cases of a cache hit when we shouldn't (see my comment on `setUpClass`)

File: torch/_inductor/codecache.py
Comment (Chillee) [reply torch/_inductor/codecache.py:331]: Probably worth adding a comment haha.

File: torch/_inductor/codecache.py
Comment (ezyang) [reply torch/_inductor/codecache.py:722]: Hmm, well I guess it didn't matter when the hints were ints but it does matter if the hints are symints. I mean, I wouldn't be opposed to just turning off this caching entirely when you're playing funny tricks with symint hints.

File: test/inductor/test_codecache.py
Comment (ezyang) [reply test/inductor/test_codecache.py:196]: resets within the test body. the main idea is that if you wobble an early count, you now no longer have to modify all the downstream counts

File: torch/_inductor/codecache.py
Comment (ezyang) [reply torch/_inductor/codecache.py:690]: Yes. You can configure artifact logging so it won't get triggered if you do `+inductor`. But I think that there is probably a non-noisy version of the logging which is acceptable on `+inductor` by default. If you are O(1) log lines per inductor compilation that is DEFINITELY acceptable

**Review by ezyang — APPROVED** [review]

File: test/inductor/test_codecache.py
Comment (ezyang) [reply test/inductor/test_codecache.py]: The "in-memory" cache is the Dynamo compiled code cache

**Pull request closed without merging**
